### ISSUE TITLE
fix: allow msg.data to be used with extract32

### ISSUE THREAD

**Issue by skellet0r** [vyperlang/vyper#2413]
### What's your issue about?

Currently `msg.data` is only allowed with the `len` and `slice` functions, it would be nice to allow `extract32` usage.

A very oversimplified example would be:

```python
def foo(_value: uint256) -> uint256:
    bar: bytes32 = extract32(msg.data, 4)
    return convert(bar, uint256)
```

A more useful scenario, particularly with [relayers](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/9fbc1d71c0ed4c68a0bc160c69df1f85e94d2d8e/contracts/metatx/MinimalForwarder.sol#L49-L51), is extracting information (like an address) that was appended to the calldata:

```python
def pretend_to_be_someone(_value: uint256) -> uint256:
    msg_sender: address = msg.sender
    if msg.sender == MAGIC_CALLER:
        msg_sender = extract32(msg.data, -20, output_type=address)
    # do something now impersonated
```

### How can it be fixed?

1. Function level validation should allow `msg.data` to be used with `extract32` as well as the previously allowed `len` and `slice` functions. [Link to source](https://github.com/vyperlang/vyper/blob/31dd77671f7cfa29af496334373a11429422226c/vyper/semantics/validation/local.py#L186-L197)

It should be changed to something like this:

```python
    def visit_Attribute(self, node):
        if node.get("value.id") == "msg" and node.attr == "data":
            parent = node.get_ancestor()
            if parent.get("func.id") not in ("slice", "len", "extract32"):
                raise SyntaxException(
                    "Invalid usage of msg.data, acceptable functions are (slice, len, extract32)",
                    node.node_source_code,
                    node.lineno,
                    node.col_offset,
                )
```

2. In the `vyper/old_codegen/expr.py` file, [here](https://github.com/vyperlang/vyper/blob/master/vyper/old_codegen/expr.py#L377-L396), my simple brain says, to just return an LLLNode which extracts the whole calldata ... but that also doesn't sound good since, calldata can be huge, and we really only want one piece of it. In which case we can extract to memory just the portion needed, from 0 to start + 32, but again also could be arbitrarily large. Alternative is to return None ? ... and then in the class [`Extract32`](https://github.com/vyperlang/vyper/blob/31dd77671f7cfa29af496334373a11429422226c/vyper/builtin_functions/functions.py#L804) identify the fact that we are interacting with calldata instead of storage or memory, to then copy the portion requested (start, start + 32) to memory and then take it from there converting to the appropriate type.

Side note: Annotation visitor `visit_Attribute` [here](https://github.com/vyperlang/vyper/blob/master/vyper/semantics/validation/annotation.py#L57-L63), could probably get scraped since in `expr.py` we can access the metadata we are storing via the `_parent` attribute of the `expr` object/argument. But that's more of a refactoring issue.

**Comment by fubuloubu** [comment]
I'd support this! `extract32` is basically `convert(slice(n, 32), bytes32)`

**Comment by skellet0r** [comment]
Yeah! Thinking some more it definitely does feel like modifying the `Extract32` class and adding a new `_calldata_element_getter`, seems like the best route

**Comment by fubuloubu** [comment]
> Yeah! Thinking some more about it, definitely does feel like modifying the `Extract32` class and adding a new `_calldata_element_getter`, seems like the best route

PRs accepted :smile:

**Comment by skellet0r** [comment]
lol, on it @fubuloubu 💪

**Comment by charles-cooper** [comment]
@skellet0r brought up a semantics issue, which is what should happen if the user calls extract32 near the boundary of calldatasize?

Here's an example:
```
@external
def foo():
    bar: bytes32 = extract32(msg.data, 0)

---- IR ---- HEAD
[with, _sub, 0, [calldataload, [mul, 32, [div, [clample, 0, [sub, [add, calldatasize, 32], 32]], 32]]]],
---- 02ca5375f45db53f39078ae99c8c743d3e86520c
[with, _sub, 0, [calldataload, [mul, 32, [div, [clample, 0, [sub, calldatasize, 32]], 32]]]],
```

In this case, calldatasize is 4. The first IR will return `method_id("foo()")` right padded with 28 zeroes. The second IR will revert. The question is which behavior we want.
